### PR TITLE
Small optimization on the datatype commit.

### PR DESCRIPTION
**Pull request by bosilca**
This patch fixes the merge of contiguous elements into larger but more
compact datatypes, and allows for contiguous elements to have thir
blocklen increasing instead of the count. The idea is to always maximize
the blocklen, aka. the contiguous part of the datatype.

Possible fix for #6932 

Signed-off-by: George Bosilca <bosilca@icl.utk.edu>

### CODE REVIEW

**Comment by bosilca** [comment]
I was able to track back the issue by looking at the datatype from issue #6932. It failed to merge into a single predefined element but the datatype was marked as contiguous, allowing the datatype engine to use one of the optimized pack/unpack function (contig_with_gaps). However, having a composed description mislead the contig_with_gaps function to wrongly interpret the shape of the datatype.

Anyway, long story short, this patch fixes the issue and add a test to prevent it from happening in the future. Once merged it should be backported to v3 and v4.

**Comment by jsquyres** [comment]
@bosilca Any particular reason not to merge this PR?

**Comment by bosilca** [comment]
It need to be backported but I don't have time to do it right now. Feel free to do so.
